File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.length();
        vector<bool> dp(n,false); 

        for(int i = 0; i < n; i++){
            for(string& ss : wordDict){
                if(i < ss.length() - 1) continue; 

                if(i == ss.length() -1 || dp[i - ss.length()]){
                    if(s.substr(i - ss.length() + 1, ss.length()) == ss){
                        dp[i] = true; 
                        break; 
                    }
                }
            }
        }

        return dp[n-1]; 
    }
};

File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        boolean[] dp = new boolean[s.length()+1]; 
        Set<String> set = new HashSet<>(); 
        for(String ss : wordDict) set.add(ss); 
        dp[0] = true; 
        for(int i = 1; i <= s.length(); i++){
            for(int j = 0; j < i; j++){
                String tmp = s.substring(j,i); 
                if(set.contains(tmp) && dp[j] == true){
                    dp[i] = true;  
                }
            }
        }

        return dp[s.length()]; 
    }
}